src/estimation/calculate_weights.py

```diff
@@ -5,6 +5,121 @@
+            including expenditure data
+        exp_col (str): An appropriate column to count n
+
+    Returns:
+        int: The number of unique references that have positive (non-negative)
+            non-Null expenditure.
+    """
+
+    # Check if any of the key cols are missing
+    cols = set(df.columns)
+    if not ("reference" in cols) & (exp_col in cols):
+        raise ValueError(f"'reference' or {exp_col} missing.")
+
+    # Count the filtered records
+    n = df["reference"].nunique()
+
+    return n
+
+
+def calculate_weighting_factor(
+    df: pd.DataFrame, cellno_dict, exp_col: str = "709"
+) -> dict:
+    """Calculate the weighting factor 'a' for each cell in the survery data
+
+    Note: A 'cell' is a group of businesses.
+
+    The calculation here is:
+
+    a = (N-o) / (n-o)
+
+    Where:
+        - N is the total number of businesses in the cell
+        - n is the number of businesses in sample for that cell
+        - o is the number of outliers in the cell
+
+    'o' is calculated in this function by summing all the `True` values
+        because `True` == 1
+
+    Args:
+        df (pd.DataFrame): The input df containing survey data
+
+    Returns:
+        weighting_factor_dict (dict): The input df with an aditional column
+
+    """
+
+    cols = set(df.columns)
+    if not ("outlier" in cols):
+        raise ValueError("The column essential 'is_outlier' is missing.")
+
+    # Convert 709 column to numeric
+    df["709"] = pd.to_numeric(df["709"], errors="coerce")
+
+    # Default a_weight = 1 for all entries
+    df["a_weight"] = 1.0
+
+    # Create filters for dataframe
+    sample_cond = df["selectiontype"] == "P"
+    status_cond = df.statusencoded.isin(["210", "211"])
+    formtype_cond = df["formtype"] == "0006"
+
+    filtered_df = df[formtype_cond & sample_cond & status_cond]
+    filtered_df = filtered_df.dropna(subset=[exp_col])
+
+    # Create small QA dataframe
+
+    qa_df = pd.DataFrame(
+        {
+            "cellnumber": [],
+            "N": [],
+            "n": [],
+            "outliers": [],
+            "a_weight": [],
+        }
+    )
+
+    # Group by cell number
+    groupd_by_cell = filtered_df.groupby("cellnumber")
+
+    # Create a dict that maps each cell to the weighting factor
+
+    for cell_number, cell_group in groupd_by_cell:
+
+        # Get N from cellno_dict
+        N = cellno_dict[cell_number]
+
+        # Get lower n
+        n = calc_lower_n(cell_group)
+
+        # Count the outliers for this group (will count all the `True` values)
+        outlier_count = cell_group["outlier"].sum()
+
+        # Calculate 'a' for this group
+        a_weight = (N - outlier_count) / (n - outlier_count)
+
+        # Put the weight into the column just for this cell number and filters
+        cell_cond = df["cellnumber"] == cell_number
+        df.loc[
+            formtype_cond & sample_cond & status_cond & cell_cond,
+            "a_weight",
+        ] = a_weight
```

Comment:
I like this! I just wonder whether we should add in the not null condition? It might be logically more sensible to have null rows with weight of 1